src/clr-core/internal/decorators/element.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { registerElementSafely } from './../utils/register';
+
+/**
+ * Class decorator factory that defines the decorated class as a custom element.
+ *
+ * ```
+ * @customElement('my-element')
+ * class MyElement {
+ *   render() {
+ *     return html``;
+ *   }
+ * }
+ * ```
+ */
+export const customElement = (tagName: string) => (classOrDescriptor: Constructor<HTMLElement> | ClassDescriptor) =>
+  typeof classOrDescriptor === 'function'
+    ? legacyCustomElement(tagName, classOrDescriptor)
+    : standardCustomElement(tagName, classOrDescriptor);
+
+// TC39 Decorators proposal
+const standardCustomElement = (tagName: string, descriptor: ClassDescriptor) => {
+  const { kind, elements } = descriptor;
+  return {
+    kind,
+    elements,
+    finisher(classDef: Constructor<HTMLElement>) {
+      registerElementSafely(tagName, classDef);
+    },
+  };
+};
+
+// Legacy TS Decorator
+const legacyCustomElement = (tagName: string, classDef: Constructor<HTMLElement>) => {
+  registerElementSafely(tagName, classDef);
+  return classDef as any;
+};
+
+// TC39 Decorators proposal
+interface ClassDescriptor {
+  kind: 'class';
+  elements: ClassElement[];
+  finisher?: <T>(classDef: Constructor<T>) => undefined | Constructor<T>;
+}
+
+interface ClassElement {
+  kind: 'field' | 'method';
+  key: PropertyKey;
+  placement: 'static' | 'prototype' | 'own';
+  initializer?: Function;
+  extras?: ClassElement[];
+  finisher?: <T>(classDef: Constructor<T>) => undefined | Constructor<T>;
+  descriptor?: PropertyDescriptor;
+}
+
+// tslint:disable-next-line:interface-over-type-literal
+export type Constructor<T> = {
+  new (...args: any[]): T;
+};
```

Comment:
This will be removed in a future version of TypeScript once they implement the type defs natively.